scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala

```diff
@@ -534,84 +520,71 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
             case (q, targs) => Type.Apply(q, targs)
           }
       }
-    case t.SINGLETON_TYPE =>
-      import s.SingletonType.Tag
-      val singleton = tpe.singletonType.get
-      def info = this.info(singleton.symbol)
-      singleton.tag match {
-        case Tag.SYMBOL =>
-          val info = this.info(singleton.symbol)
-          if (info.kind.isParameter || info.isVal || info.kind.isObject) {
-            Type.Singleton(toTermRef(info))
-          } else {
-            val tpe = info.typ
-            tpe.tag match {
-              case t.METHOD_TYPE =>
-                val ret = tpe.methodType.get.returnType.get
-                toType(ret)
-              case _ =>
-                fail(tpe)
-            }
-          }
-        case Tag.THIS =>
-          Type.Select(
-            Term.This(Name.Anonymous()),
-            Type.Name(info.name)
-          )
-        case Tag.SUPER =>
-          // TODO: prefix
-          Type.Select(
-            Term.Super(Name.Anonymous(), Name.Anonymous()),
-            Type.Name(info.name)
-          )
-        case _ =>
-          toType(tpe.widen)
+    case s.SingleType(_, symbol) =>
+      val info = this.info(symbol)
+      if (info.kind.isParameter || info.isVal || info.kind.isObject) {
+        Type.Singleton(toTermRef(info))
+      } else {
+        info.signature match {
+          case s.MethodSignature(_, _, returnType) =>
+            toType(returnType)
+          case els =>
+            fail(els)
+        }
       }
-    case t.EXISTENTIAL_TYPE =>
-      val existential = tpe.existentialType.get
-      withPlaceholders(existential.typeParameters) { () =>
-        toType(existential.tpe.get)
+    case s.ThisType(symbol) =>
+      Type.Select(
+        Term.This(Name.Anonymous()),
+        Type.Name(this.info(symbol).name)
+      )
+    case s.SuperType(prefix @ _, symbol) =>
+      // TODO: prefix
+      Type.Select(
+        Term.Super(Name.Anonymous(), Name.Anonymous()),
+        Type.Name(this.info(symbol).name)
+      )
+    case s.ConstantType(_) =>
+      toType(tpe.widen)
+    case s.ExistentialType(underlying, Some(declarations)) =>
+      withHardlinks(declarations.hardlinks) { () =>
```

Comment:
How did you find dealing with hardlinks?

Reply:
It's been great so far 😊 Removed a lot of `???`/WTFs in the old implementation. I currently keep a mutable scope for hardlink lookups, I could make it immutable but that would require adding a `(hardlinks: Map[String, SymbolInformation)` parameter everywhere for questionable benefit.